test/unit/extension_video_sensing.js

```diff
@@ -0,0 +1,408 @@
+const {createReadStream} = require('fs');
+const {join} = require('path');
+
+const {PNG} = require('pngjs');
+const {test} = require('tap');
+
+const VideoSensing = require('../../src/extensions/scratch3_video_sensing/index.js');
+const VideoMotion = require('../../src/extensions/scratch3_video_sensing/library.js');
+
+/**
+ * Prefix to the mock frame images used to test the video sensing extension.
+ * @type {string}
+ */
+const pngPrefix = 'extension_video_sensing_';
+
+/**
+ * Map of frame keys to the image filenames appended to the pngPrefix.
+ * @type {object}
+ */
+const framesMap = {
+    center: 'center',
+    left: 'left-5',
+    left2: 'left-10',
+    down: 'down-10'
+};
+
+/**
+ * Asynchronously read a png file and copy its pixel data into a typed array
+ * VideoMotion will accept.
+ * @param {string} name - partial filename to read
+ * @returns {Promise.<Uint32Array>} pixel data of the image
+ */
+const readPNG = name => (
+    new Promise((resolve, reject) => {
+        const png = new PNG();
+        createReadStream(join(__dirname, `${pngPrefix}${name}.png`))
+            .pipe(png)
+            .on('parsed', () => {
+                // Copy the RGBA pixel values into a separate typed array and
+                // cast the array to Uint32, the array format VideoMotion takes.
+                resolve(new Uint32Array(new Uint8ClampedArray(png.data).buffer));
```

Comment:
Is this different from `resolve(new Uint32Array(png.data.buffer))`? My reading of the `pngjs` code suggests that `png.data` is a `Buffer`, which means that in Node 4.x+ it's also a `Uint8Array` already, so it should just be a view on the same actual data...?

Reply:
It's the use of `new Uint8ClampedArray(png.data)` to first make a copy as a object ownership safety in case `png.data` was externally modified. `Uint8Array` would be fine, using `Uint8ClampedArray` since its more canonical for pixel data manipulation.

I can toss the copy. I suppose its a personal default behaviour of mine for this kind of scenario (loading an image) when high performance isn't a priority, like in a test.

Reply:
Oh, I see. That's cool - no objection now that I understand what you're doing, especially in a test like you said :)